### ISSUE TITLE
IDE not indexing generated source

### ISSUE THREAD

**Issue by davidjwiner** [google/ksp#37]
Looks like that kapt has an IDE plugin to achieve this: KaptProjectResolverExtension.kt

We either need to copy it over, or simply tell users to mark generated source root themselves in the preview.

**Comment by ting-yuan** [comment]
Deferring again to 2021Q1.

**Comment by seongug-jung** [comment]
```kotlin
buildTypes {
    getByName("debug") {
        sourceSets {
            getByName("main") {
                java.srcDir(File("build/generated/ksp/debug/kotlin"))
            }
        }
    }
    getByName("release") {
        sourceSets {
            getByName("main") {
                java.srcDir(File("build/generated/ksp/release/kotlin"))
            }
        }
    }
}
```
for now.

**Comment by ZeroBrain** [comment]
```kotlin
buildTypes {
    getByName("debug") {
        isMinifyEnabled = false
        proguardFiles(getDefaultProguardFile("proguard-android-optimize.txt"), "proguard-rules.pro")
        buildConfigField("boolean", "DEBUG_BUILD", "true")
    }

    getByName("release") {
        isMinifyEnabled = false
        proguardFiles(getDefaultProguardFile("proguard-android-optimize.txt"), "proguard-rules.pro")
        buildConfigField("boolean", "DEBUG_BUILD", "false")
    }
}
```
I am using above script in build.gradle.kts and working as well for now.

**Comment by keyboardr** [comment]
As a slight modification to the above, here's an approach that's a little more compact and flexible:
```kotlin
    applicationVariants.all {
        val variantName = name
        sourceSets {
            getByName("main") {
                java.srcDir(File("build/generated/ksp/$variantName/kotlin"))
            }
        }
    }
```

**Comment by gmazzo** [comment]
We use this logic in our tooling to property support all extension types:
```kotlin
    fun Project.setupKSPOutputsAsGeneratedFolders() = afterEvaluate {
        plugins.withId("com.google.devtools.ksp") {
            the<BaseExtension>().variants.configureEach variant@{
                registerJavaGeneratingTask(
                    tasks.named("ksp${this@variant.name.capitalize(Locale.ROOT)}Kotlin"),
                    layout.buildDirectory.dir("generated/ksp/debug/kotlin").get().asFile
                )
            }
        }
    }

    val BaseExtension.variants: DomainObjectSet<out BaseVariant>
        get() = when (this) {
            is AppExtension -> applicationVariants
            is LibraryExtension -> libraryVariants
            is TestExtension -> applicationVariants
            else -> error("unsupported module type: $this")
        }
```

**Comment by naingaungluu-codigo** [comment]
For folks who want to set kotlin sourceSets, we can configure it this way 
```groovy
android.applicationVariants.all { variant ->
    kotlin.sourceSets {
        def name = variant.name
        getByName(name) {
            kotlin.srcDir("build/generated/ksp/$name/kotlin")
        }
    }
}
```

**Comment by apkelly** [comment]
This worked for us in an android library module.

    androidComponents.onVariants {variant ->
        kotlin.sourceSets.findByName(variant.name)?.kotlin?.srcDirs(
            file("$buildDir/generated/ksp/${variant.name}/kotlin")
        )
    }

**Comment by eugene-krivobokov** [comment]
I'd like to warn about issues in previously mentioned options.

```kotlin
    applicationVariants.all {
        val variantName = name
        sourceSets {
            getByName("main") {
                java.srcDir(File("build/generated/ksp/$variantName/kotlin"))
            }
        }
    }
```
`sourceSets` is not variant aware.
This code adds generated code from all build variants to the main source set. 
E.g. generated code for "debug" and "release" build variants will be in "main" source set.
Print out source sets to see the issues.
It goes against the idea of build variants. This is probably not what you want.

```kotlin
configure<BaseExtension> {
    buildVariants.configureEach {
        sourceSets {
            named(name).configure {
                java.srcDir("build/generated/ksp/$buildVariantName/kotlin")
            }
        }
    }
}
```

This alternative has another issue.
It adds generated code as input and breaks build cache.
I'll show this on example.

First clean build:

```
./gradlew :module:clean  :module:kspReleaseKotlin -Dorg.gradle.caching.debug=true

Appending input file fingerprints for 'source' to build cache key: xxx - RELATIVE_PATH{/project/module/src/main/java/.../Clazz.kt' / ...}
```

There were only sources as input for this task.
Now, lets run again:

```
./gradlew  :module:kspReleaseKotlin -Dorg.gradle.caching.debug=true

Appending input file fingerprints for 'source' to build cache key: yyy - RELATIVE_PATH{/project/module/build/generated/ksp/release/kotlin/<generated>, /project/module/src/main/java/.../Clazz.kt' / ...}
```

Here we see generated files as and addition to inputs. It causes a cache miss. The output is not "stable".

**Comment by phucynwa** [comment]
> For android folks who want to set kotlin sourceSets, we can configure it this way
> 
> ```groovy
> android.applicationVariants.all { variant ->
>     kotlin.sourceSets {
>         def name = variant.name
>         getByName(name) {
>             kotlin.srcDir("build/generated/ksp/$name/kotlin")
>         }
>     }
> }
> ```

For build.gradle.kts
```kotlin
    android.applicationVariants.all {
        kotlin {
            sourceSets {
                getByName(name) {
                    kotlin.srcDir("build/generated/ksp/$name/kotlin")
                }
            }
        }
    }
```

**Comment by lukaszkalnik** [comment]
@eugene-krivobokov can you suggest correct solution for Android when using build variants? Thank you 🙏

**Comment by leinardi** [comment]
After upgrading to Android Studio Electric Eel the following solution didn't work anymore:

```groovy
android {
    libraryVariants.all {
        kotlin.sourceSets {
            getByName(name) {
                kotlin.srcDir("build/generated/ksp/${name}/kotlin")
            }
        }
    }
}
```

But this one seems to work fine:
```groovy
android {
    kotlin {
        sourceSets {
            debug {
                kotlin.srcDir("build/generated/ksp/debug/kotlin")
            }
            release {
                kotlin.srcDir("build/generated/ksp/release/kotlin")
            }
        }
    }
}
```

**Comment by Jacob3075** [comment]
None of the above solutions work on `Android Studio Flamingo | 2022.2.1 Canary 9`
kotlin version: `1.7.20`
ksp: `1.7.20-1.0.8`
AGP: `8.0.0-alpha09`
Gradle: `7.6`

The solution that leinardi gave was working around canary 4 but I don't remember in which version it stopped.

**Comment by Raven-L** [comment]
```
    @Suppress("UnstableApiUsage")
    sourceSets.configureEach {
        kotlin.srcDir("$buildDir/generated/ksp/$name/kotlin/")
    }
```
it works well for me.
[come from](https://www.kodeco.com/33148161-write-a-symbol-processor-with-kotlin-symbol-processing)

**Comment by ting-yuan** [comment]
Fixed in e81d01f80e8da40542f22e8cbeec724aa22d6062 and will be available in 1.0.9. There should no longer workarounds be needed.

**Comment by eygraber** [comment]
@ting-yuan using 1.0.11 it looks like the following is still required:

```kotlin
sourceSets.named("commonMain").configure {
    kotlin.srcDir("build/generated/ksp/metadata/commonMain/kotlin")
}
```

Is that expected?
